### PR TITLE
evaluation/retrieval.py: multicaption + video2text support

### PR DESCRIPTION
**Pull request by danielmend**

### CODE REVIEW

File: src/aggregation/mean.py [hunk 1]
```diff
@@ -5,9 +5,9 @@
...
 
 from torch import nn
-
+import torch.nn.functional as F
 class Mean(nn.Module):
   def __init__(self):
     super().__init__()
   def forward(self, x):
-    return x.mean(axis=-2) # assumes shape always ends with (..., n_frames, embed_dim)
+    return F.normalize(x.mean(axis=-2), dim=-1) # assumes shape always ends with (..., n_frames, embed_dim)
```
Comment (iejMac) [inline src/aggregation/mean.py:13]: I think let's keep it as is right now and do a large 'normalizing' themed refactoring next, it's good you bring it up.

File: src/evaluation/retrieval.py [hunk 2]
```diff
@@ -1,80 +1,87 @@
...
-    return val_metrics
+    return val_metrics,
 
+def get_metrics(video_features, text_features, logit_scale):
+    ''' 
+    Assumptions for this eval:
 
-def get_metrics(video_features, text_features, ground_truth, logit_scale):
-    metrics = {}
+    - len(text_features) % len(video_features) == 0, i.e. there's a constant number
```
Comment (iejMac) [inline src/evaluation/retrieval.py:48]: this assumption isn't true for MSVD (a common benchmark) for example and I'm not sure if we should have this...

File: src/evaluation/retrieval.py [hunk 3]
```diff
@@ -1,80 +1,106 @@
...
             for cap in batch["text"]:
-                if multicaption:
-                    for c in cap.split(";"): # multiple captions separated by ;
-                        toks.append(open_clip.tokenize(c))
-                        ground_truth.append(samp)
-                else:
-                    toks.append(open_clip.tokenize(cap))
-                    ground_truth.append(samp)
-                samp += 1
+              caps_list = cap.split(';')
```
Comment (iejMac) [inline src/evaluation/retrieval.py:21]: why aren't we checking for the 'multicaption' param? maybe some datasets have one caption but the caption has ';' in it?

File: src/evaluation/retrieval.py [hunk 4]
```diff
@@ -1,80 +1,106 @@
...
+              if len(caps_list) > max_txt_len:
```
Comment (iejMac) [inline src/evaluation/retrieval.py:22]: I hate if's so can we do something like ```max_txt_len = max(max_txt_len, len(caps_list)``` or at least use a ternary op

File: src/evaluation/retrieval.py [hunk 5]
```diff
@@ -1,80 +1,106 @@
...
+  out = []
+  for mat in text_features:
+    padded = torch.zeros(size=(max_txt_len, dim_model))
+    padded[0:len(mat), :] = mat
+    
+    out.append(padded)
 
-def get_metrics(video_features, text_features, ground_truth, logit_scale):
-    metrics = {}
+  return torch.stack(out)
```
Comment (iejMac) [inline src/evaluation/retrieval.py:56]: you want to be explicit about what your functions do. This is "zero_pad_text_features" so the output should be the input but zero-padded, there's no mention of stacking so I'd just put that part back into the function that calls this.

File: src/evaluation/retrieval.py [hunk 6]
```diff
@@ -1,80 +1,106 @@
...
-        preds = torch.where(ranking == ground_truth)[1]
-        preds = preds.detach().cpu().numpy()
-        metrics[f"{name}_mean_rank"] = preds.mean() + 1
-        metrics[f"{name}_median_rank"] = np.floor(np.median(preds)) + 1
+        argsorted = torch.argsort(logit, dim=-1, descending=True)
+        ranks_on_diagonals = torch.argsort(argsorted, dim=-1, descending=False)
+        
+        rankings = torch.diagonal(ranks_on_diagonals, dim1=-1, dim2=-2).flatten().float()
+
+        if name == 'text_to_video':
```
Comment (iejMac) [inline src/evaluation/retrieval.py:96]: can we do without this if?

File: src/evaluation/retrieval.py (hunk 5 above)
Comment (danielmend) [reply src/evaluation/retrieval.py:56]: the thing is that the input is a torch tensor but `out` is a list. torch stacking `out` make it so `out` is the same type as the `text_features`, i.e. it actually is just a zero padded tensor.

Comment (iejMac) [reply src/evaluation/retrieval.py:56]: https://github.com/danielmend/temporal-embedding-aggregation/blob/2a3bc2066f383acc386450dd7a7c8db17ec7dd9b/src/evaluation/retrieval.py#L38
input is all_text_features which is a list since we append to it

Comment (danielmend) [reply src/evaluation/retrieval.py:56]: i dont even remember how my own code works 😂  ok on it.

Comment (iejMac) [reply src/evaluation/retrieval.py:56]: all good, I'd just wrap the zero_pad_text_features in torch.stack instead of wrapping out in torch.stack
